### PR TITLE
Extract node module resolution from JavascriptImport into a separate module

### LINKED ISSUE #921: import `./foo` doesn't convert to `./foo.js`
Need to use node resolution algorithm for resolving relative modules - `./foo` could be `./foo.js` or `./foo/index.js`. Affects polyserve.

cc @usergenic

### PR DESCRIPTION
**Pull request by justinfagnani**
Fixes #921 

 - [x] CHANGELOG.md has been updated

### CODE REVIEW

File: src/javascript/javascript-import-scanner.ts [hunk 1]
```diff
@@ -254,32 +236,34 @@ export class JavaScriptImportScanner implements JavaScriptScanner {
...
+      try {
+        // TODO (justinfagnani): we need to pass packageName and componentDir,
+        // or maybe the UrlResolver, to get correct relative URLs for
+        // named imports from the root package to its dependencies
+        return resolve(specifier, documentPath);
+      } catch (e) {
+        if (!isPathSpecifier(specifier)) {
+          // Don't warn if the specifier was already a path, even though we do
+          // resolve paths, because maybe the user is serving it some other way.
+          console.warn(`Could not resolve module specifier "${specifier}"`);
```
Comment (rictic) [inline src/javascript/javascript-import-scanner.ts]: The comment says that we don't warn, but the code does warn. Is this debug code

IMO we should not log to console, we should emit a Warning or throw an Error to express a problem.

File: src/javascript/resolve-specifier-node.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+              return packageJson;
+            },
+      });
+
+      let relativeSpecifierUrl =
+          relative(dirname(documentPath), resolvedSpecifier) as FileRelativeUrl;
+
+      if (componentInfo !== undefined) {
+        const {packageName, rootDir, componentDir} = componentInfo;
+        console.log(
```
Comment (rictic) [inline src/javascript/resolve-specifier-node.ts]: console log

File: src/javascript/resolve-specifier-node.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * Resolves module specifiers using node module resolution.
+ *
+ * Full URLs - those parsable by the WHATWG URL spec - are returned as-is.
+ * Absolute and relative paths are resolved, even though they are valid
+ * HTML-spec module specifiers, because node resolution supports directories
+ * and omitting extensions. If a specifier doesn't resolve, it's returned as-is.
+ */
+export const resolve =
+    (specifier: string, documentPath: string, componentInfo?: {
```
Comment (rictic) [inline src/javascript/resolve-specifier-node.ts]: Document what `componentInfo` means?

File: src/javascript/javascript-import-scanner.ts (hunk 1 above)
Comment (aomarks) [reply src/javascript/javascript-import-scanner.ts]: The comment applies to the if statement (maybe clearer to say "Only warn if the specifier was a name")

**Comment by justinfagnani** [comment]
PTAL

File: src/javascript/resolve-specifier-node.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+ *
+ * Full URLs - those parsable by the WHATWG URL spec - are returned as-is.
+ * Absolute and relative paths are resolved, even though they are valid
+ * HTML-spec module specifiers, because node resolution supports directories
+ * and omitting extensions. If a specifier doesn't resolve, it's returned as-is.
+ *
+ * @param componentInfo An object describing a "component-style" URL layout. In
+ *   this layout, cross-package URLs reach out of the package directory to
+ *   sibling packages, rather than into the component directory. When given,
+ *   this parameter causes relative paths to be returns for this style.
```
Comment (aomarks) [inline src/javascript/resolve-specifier-node.ts]: to be returned

File: src/javascript/javascript-import-scanner.ts [hunk 5]
```diff
@@ -254,32 +236,30 @@ export class JavaScriptImportScanner implements JavaScriptScanner {
...
-      relativeSpecifierUrl =
-          relativeSpecifierUrl.replace(/\\/g, '/') as FileRelativeUrl;
+    if (this.moduleResolution === 'node') {
+      try {
+        // TODO (justinfagnani): we need to pass packageName and componentDir,
+        // or maybe the UrlResolver, to get correct relative URLs for
+        // named imports from the root package to its dependencies
+        return resolve(specifier, documentPath);
+      } catch (e) {
+        // If the specifier was a name, we'll emit a warning below.
```
Comment (rictic) [inline src/javascript/javascript-import-scanner.ts]: The warning emitted below though is too specific, it's just for when there's no module resolution algorithm.

File: src/javascript/resolve-specifier-node.ts (hunk 4 above)
Comment (rictic) [reply src/javascript/resolve-specifier-node.ts]: This has really heavy overlap with PackageUrlResolver. Can we unify?

**Pull request merged**